### ISSUE TITLE
Status Code: 502 returned intermittently by v2 instances

### ISSUE THREAD

**Issue by henokgetachew** [medic/medic-webapp#3043]
We've been seeing 502 status code returned intermittently when medic-bot tries to access our v2 instances.  These outages are not permanent and therefore haven't been debugged thoroughly.  This issue is a placeholder where we compile and investigate these intermittent outages that happen every once in a while.

**Comment by henokgetachew** [comment]
Investigation after one such failure:
API goes down with an uncaught exception shown below at 14:01.  MedicBot does the check at 14:03 and reports failure.  API comes back up 14:05 right before MedicBot's second schedule which would have been 14:06.

API logs:
```
**2017-02-28T14:01:00.063Z** - error: UNCAUGHT EXCEPTION!                                                                                                                                                               
                                                                                                                                                                                                                    
2017-02-28T14:01:00.064Z - error: Error: write after end                                                                                                                                                            
    at ServerResponse.OutgoingMessage.write (_http_outgoing.js:413:15)                                                                                                                                              
    at prepareResponse (/srv/storage/gardener/data/working_dir/aHR0cDovL2xvY2FsaG9zdDo1OTg0L21lZGljL19kZXNpZ24vbWVkaWMvbWVkaWMtYXBp/node_modules/medic-api/handlers/changes.js:139:12)                              
    at /srv/storage/gardener/data/working_dir/aHR0cDovL2xvY2FsaG9zdDo1OTg0L21lZGljL19kZXNpZ24vbWVkaWMvbWVkaWMtYXBp/node_modules/medic-api/handlers/changes.js:168:7                                                 
    at Request._callback (/srv/storage/gardener/data/working_dir/aHR0cDovL2xvY2FsaG9zdDo1OTg0L21lZGljL19kZXNpZ24vbWVkaWMvbWVkaWMtYXBp/node_modules/medic-api/node_modules/nano/lib/nano.js:217:16)                  
    at Request.self.callback (/srv/storage/gardener/data/working_dir/aHR0cDovL2xvY2FsaG9zdDo1OTg0L21lZGljL19kZXNpZ24vbWVkaWMvbWVkaWMtYXBp/node_modules/medic-api/node_modules/request/request.js:186:22)            
    at Request.emit (events.js:110:17)                                                                                                                                                                              
    at Request.<anonymous> (/srv/storage/gardener/data/working_dir/aHR0cDovL2xvY2FsaG9zdDo1OTg0L21lZGljL19kZXNpZ24vbWVkaWMvbWVkaWMtYXBp/node_modules/medic-api/node_modules/request/request.js:1081:10)             
    at Request.emit (events.js:107:17)                                                                                                                                                                              
    at IncomingMessage.<anonymous> (/srv/storage/gardener/data/working_dir/aHR0cDovL2xvY2FsaG9zdDo1OTg0L21lZGljL19kZXNpZ24vbWVkaWMvbWVkaWMtYXBp/node_modules/medic-api/node_modules/request/request.js:1001:12)     
    at IncomingMessage.g (events.js:199:16)                                                                                                                                                                         
                                                                                                                                                                                                                    
2017-02-28T14:01:00.068Z - error: Server error: { [Error: write after end]                                                                                                                                          
  domainEmitter:                                                                                                                                                                                                    
   { domain:                                                                                                                                                                                                   
      { domain: null,                                                                                                                                                                                               
        _events: [Object],                                                                                                                                                                                     
        _maxListeners: undefined,                                                                                                                                                                                   
        members: [] },                                                                                                                                                                                         
     _events: { prefinish: [Function: resOnFinish] },                                                                                                                                                               
     _maxListeners: undefined,                                                                                                                                                                                 
     output: [],                                                                                                                                                                                                    
     outputEncodings: [],                                                                                                                                                                                      
     outputCallbacks: [],                                                                                                                                                                                           
     writable: true,                                                                                                                                                                                           
     _last: true,                                                                                                                                                                                                   
     chunkedEncoding: false,                                                                                                                                                                                   
     shouldKeepAlive: false,                                                                                                                                                                                        
     useChunkedEncodingByDefault: false,                                  
     sendDate: true,                                                                                                                                                                                                
     _removedHeader: {},                             
     _hasBody: true,                                                                                                                                                                                                
     _trailer: '',                                   
     finished: true,                                                                                                                                                                                                
     _hangupClose: false,                            
     _headerSent: true,                                                                                                                                                                                             
     socket: null,                                   
     connection: null,                                                                                                                                                                                              
     _header: 'HTTP/1.1 200 OK\r\nX-Powered-By: Express\r\nContent-Type: application/json\r\nDate: Tue, 28 Feb 2017 13:57:25 GMT\r\nConnection: close\r\n\r\n',
     _headers:                                                                                                                                                                                                      
      { 'x-powered-by': 'Express',                                                                                                                             
        'content-type': 'application/json' },                                                                                                                                                                       
     _headerNames:                                                                                                                                             
      { 'x-powered-by': 'X-Powered-By',                                                                                                                                                                             
        'content-type': 'Content-Type' },                    
     req:                                                                                                                                                                                                           
      { _readableState: [Object],                                                                                                                                                                                   
        readable: false,                                                                                                                                                                                            
        domain: [Object],                                                                                                                                                                                           
        _events: [Object],                                                                                                                                                                                          
        _maxListeners: undefined,                                                                                                                                                                                   
        socket: [Object],                                                                                                                                                                                           
        connection: [Object],                                                                                                                                                                                       
        httpVersionMajor: 1,                                                                                                                                                                                        
        httpVersionMinor: 0,                                                                                                                                                                                        
        httpVersion: '1.0',                                                                                                                                                                                         
        complete: true,                                                                                                                                                                                             
        headers: [Object],                                                                                                                                                                                          
        rawHeaders: [Object],                                                                                                                                                                                       
        trailers: {},                                                                                                                                                                                               
        rawTrailers: [],                                                                                                                                                                                            
        _pendings: [],                                                                                                                                                                                              
        _pendingIndex: 0,                                                                                                                                                                                           
        upgrade: false,                                                                                                                                                                                             
        url: '/medic/_changes?style=all_docs&feed=longpoll&heartbeat=10000&filter=_doc_ids&since=3149205&limit=95',                                                                                                 
        method: 'POST',                                                                                                                                                                                             
        statusCode: null,                                                                                                                                                                                           
        statusMessage: null,                                                                                                                                                                                        
        client: [Object],                                                                                                                                                                                           
        _consuming: true,                                                                                                                                                                                           
        _dumped: false,                                                                                                                                                                                        
        next: [Function: next],                                                                                                                                                                                     
        baseUrl: '',                                                                                                                                                                                           
        originalUrl: '/medic/_changes?style=all_docs&feed=longpoll&heartbeat=10000&filter=_doc_ids&since=3149205&limit=95',                                                                                         
        _parsedUrl: [Object],                                                                                                                                                                                  
        params: {},                                                                                                                                                                                                 
        query: [Object],                                                                                                                                                                                       
        res: [Circular],                                                                                                                                                                                            
        _startAt: [Object],                                                                                                                                                                                    
        _startTime: Tue Feb 28 2017 13:57:14 GMT+0000 (UTC),                                                                                                                                                        
        _remoteAddress: '127.0.0.1',                                                                                                                                                                           
        route: [Object],                                                                                                                                                                                            
        body: [Object],                                                                                                                                                                                        
        _body: true,                                                                                                                                                                                                
        length: undefined,                                                                                                                                     
        read: [Function] },                                                                                                                                                                                         
     locals: {},                                                                                                                                               
     _startAt: undefined,                                                                                                                                                                                           
     _startTime: undefined,                                                                                                                                    
     statusMessage: 'OK',                                                                                                                                                                                           
     statusCode: 200 },                                                                                                                                        
  domain:                                                                                                                                                                                                           
   { domain: null,                                                                                                                                             
     _events: { error: [Function] },                                                                                                                                                                                
     _maxListeners: undefined,                                                                                                                                 
     members: [] },                                                                                                                                                                                                 
  domainThrown: false }                                                                                                                                        
                                                                                                                                                                                                                    
                                                                                                                                                                                                                    
2017-02-28T14:01:00.068Z - error: UNCAUGHT EXCEPTION!                                                                                                                                                               
                                                                                                                                                                                                                    
2017-02-28T14:01:00.069Z - error: Error: Can't set headers after they are sent.                                                                                                                                     
    at ServerResponse.OutgoingMessage.setHeader (_http_outgoing.js:335:11)                                                                                                                                          
    at ServerResponse.res.writeHead (/srv/storage/gardener/data/working_dir/aHR0cDovL2xvY2FsaG9zdDo1OTg0L21lZGljL19kZXNpZ24vbWVkaWMvbWVkaWMtYXBp/node_modules/medic-api/server.js:509:9)                            
    at Object.module.exports.serverError (/srv/storage/gardener/data/working_dir/aHR0cDovL2xvY2FsaG9zdDo1OTg0L21lZGljL19kZXNpZ24vbWVkaWMvbWVkaWMtYXBp/node_modules/medic-api/server-utils.js:35:9)                  
    at Domain.<anonymous> (/srv/storage/gardener/data/working_dir/aHR0cDovL2xvY2FsaG9zdDo1OTg0L21lZGljL19kZXNpZ24vbWVkaWMvbWVkaWMtYXBp/node_modules/medic-api/server.js:66:17)                                      
    at Domain.emit (events.js:107:17)                                                                                                                                                                               
    at ServerResponse.emit (events.js:83:19)                                                                                                                                                                        
    at _http_outgoing.js:415:12                                                                                                                                                                                     
    at process._tickDomainCallback (node.js:381:11)                                                                                                                                                                 
                                                                                                                                                                                                                    
2017-02-28T14:01:00.069Z - error: Server error: { [Error: Can't set headers after they are sent.]                                                                                                                   
  domain:                                                                                                                                                                                                           
   { domain: null,                                                                                                                                                                                                  
     _events: { error: [Function] },                                                                                                                                                                                
     _maxListeners: undefined,                                                                                                                                                                                      
     members: [] },                                                                                                                                                                                                 
  domainThrown: true }                                                                                                                                                                                              
                                                                                                                                                                                                                    
2017-02-28T14:01:00.069Z - error: UNCAUGHT EXCEPTION!                                                                                                                                                               
                                                                                                                                                                                                                    
2017-02-28T14:01:00.070Z - error: Error: Can't set headers after they are sent.                                                                                                                                     
    at ServerResponse.OutgoingMessage.setHeader (_http_outgoing.js:335:11)                                                                                                                                          
    at ServerResponse.res.writeHead (/srv/storage/gardener/data/working_dir/aHR0cDovL2xvY2FsaG9zdDo1OTg0L21lZGljL19kZXNpZ24vbWVkaWMvbWVkaWMtYXBp/node_modules/medic-api/server.js:509:9)                            
    at Object.module.exports.serverError (/srv/storage/gardener/data/working_dir/aHR0cDovL2xvY2FsaG9zdDo1OTg0L21lZGljL19kZXNpZ24vbWVkaWMvbWVkaWMtYXBp/node_modules/medic-api/server-utils.js:35:9)                  
    at Domain.<anonymous> (/srv/storage/gardener/data/working_dir/aHR0cDovL2xvY2FsaG9zdDo1OTg0L21lZGljL19kZXNpZ24vbWVkaWMvbWVkaWMtYXBp/node_modules/medic-api/server.js:66:17)                                      
    at Domain.emit (events.js:107:17)                                                                                                                                                                          
    at Domain.errorHandler [as _errorHandler] (domain.js:102:19)                                                                                                                                                    
    at process._fatalException (node.js:233:33)                                                                                                                                                                
                                                                                                                                                                                                                    
2017-02-28T14:01:00.070Z - error: Server error: { [Error: Can't set headers after they are sent.]                                                                                                              
  domain:                                                                                                                                                                                                           
   { domain: null,                                                                                                                                                                                             
     _events: { error: [Function] },                                                                                                                                                                                
     _maxListeners: undefined,                                                                                                                                                                                 
     members: [] },                                                                                                                                                                                                 
  domainThrown: true }                                                                                                                                                                                         
                                                                                                                                                                                                                    
2017-02-28T14:01:00.071Z - error: _http_outgoing.js:335                                                                                                                                                        
    throw new Error('Can\'t set headers after they are sent.');                                                                                                                                                     
          ^                                                                                                                                                                                       
Error: Can't set headers after they are sent.                                                                                                                                                                       
    at ServerResponse.OutgoingMessage.setHeader (_http_outgoing.js:335:11)                                                                                                                        
    at ServerResponse.res.writeHead (/srv/storage/gardener/data/working_dir/aHR0cDovL2xvY2FsaG9zdDo1OTg0L21lZGljL19kZXNpZ24vbWVkaWMvbWVkaWMtYXBp/node_modules/medic-api/server.js:509:9)                            
    at Object.module.exports.serverError (/srv/storage/gardener/data/working_dir/aHR0cDovL2xvY2FsaG9zdDo1OTg0L21lZGljL19kZXNpZ24vbWVkaWMvbWVkaWMtYXBp/node_modules/medic-api/server-utils.js:35:9)
    at Domain.<anonymous> (/srv/storage/gardener/data/working_dir/aHR0cDovL2xvY2FsaG9zdDo1OTg0L21lZGljL19kZXNpZ24vbWVkaWMvbWVkaWMtYXBp/node_modules/medic-api/server.js:66:17)                                      
    at Domain.emit (events.js:107:17)                                                                                                                                                             
    at ServerResponse.emit (events.js:83:19)                                                                                                                                                                        
    at _http_outgoing.js:415:12                                                                                                                                                                                     
    at process._tickDomainCallback (node.js:381:11)                                                                                                                                                                 
                                                                                                                                                                                                                    
**2017-02-28T14:05:31.179Z** - info: Database migrations completed successfully                                                                                                                                         
                                                                                                                                                                                                                    
2017-02-28T14:05:31.319Z - info: Medic API listening on port 5988
```

**Comment by garethbowen** [comment]
This is a dupe of https://github.com/medic/medic-webapp/issues/3111 and fixed in soon-to-be released 2.10

**Comment by henokgetachew** [comment]
That might very well mean that our intermittent 502's have gone for good. We would have been wondering why they disappeared after the update of 2.10.
